### PR TITLE
Run time comparisons taking account of possible clock skew

### LINKED ISSUE #20: Review: Disable unreliable check between current times of different systems #17
Given a bug found in our verifier, this change was performed:

https://github.com/adorsys/sd-jwt/pull/17/files

Revise the question and check if it makes sense within the specs.

### PR DESCRIPTION
**Pull request by IngridPuppet**
Closes #20 

In a nutshell:
- Completely remove `iat` checks beyond emptying the corresponding method definition.
- Introduce a configurable leeway for `nbf` and `exp` checks.

### CODE REVIEW

File: src/main/java/de/adorsys/sdjwt/IssuerSignedJwtVerificationOpts.java [hunk 1]
```diff
@@ -7,75 +7,41 @@
...
-            this.validateExpirationClaim = validateExpirationClaim;
-            return this;
-        }
-
-        public Builder withValidateNotBeforeClaim(boolean validateNotBeforeClaim) {
-            this.validateNotBeforeClaim = validateNotBeforeClaim;
-            return this;
-        }
-
         public IssuerSignedJwtVerificationOpts build() {
```
Comment (Ogenbertrand) [inline src/main/java/de/adorsys/sdjwt/IssuerSignedJwtVerificationOpts.java:39]: There’s a potential null pointer exception if JWSVerifier verifier is not set in the Builder?

File: src/test/java/de/adorsys/sdjwt/TimeClaimVerifierTest.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            return CURRENT_TIMESTAMP;
+        }
+    }
+
+    @Test
+    public void testVerifyExpClaimExpired() {
+        ObjectNode payload = SdJwtUtils.mapper.createObjectNode();
+        payload.put("exp", CURRENT_TIMESTAMP - 100); // Expired 100 seconds ago
+
+        assertThrows(SdJwtVerificationException.class, () -> timeClaimVerifier.verifyExpClaim(payload));
```
Comment (Ogenbertrand) [inline src/test/java/de/adorsys/sdjwt/TimeClaimVerifierTest.java]: instead of just using assertThrows, store the exception in a variable to make it clear what exception is expected. what do you think @IngridPuppet

File: src/test/java/de/adorsys/sdjwt/TimeClaimVerifierTest.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    @Test
+    public void testVerifyExpClaimValid() throws SdJwtVerificationException {
```
Comment (Ogenbertrand) [inline src/test/java/de/adorsys/sdjwt/TimeClaimVerifierTest.java:43]: Is this method suppose to throw any exception?, if yes why not throwing an the exception to this method!

File: src/test/java/de/adorsys/sdjwt/TimeClaimVerifierTest.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    @Test
+    public void testVerifyNotBeforeClaimNotYetValid() {
+        ObjectNode payload = SdJwtUtils.mapper.createObjectNode();
+        payload.put("nbf", CURRENT_TIMESTAMP + 100); // Not valid for another 100 seconds
+
+        assertThrows(SdJwtVerificationException.class, () -> timeClaimVerifier.verifyNotBeforeClaim(payload));
+    }
+
+    @Test
+    public void testVerifyNotBeforeClaimValid() throws SdJwtVerificationException {
```
Comment (Ogenbertrand) [inline src/test/java/de/adorsys/sdjwt/TimeClaimVerifierTest.java:71]: same here

**Review by Ogenbertrand — CHANGES_REQUESTED** [review]

File: src/main/java/de/adorsys/sdjwt/IssuerSignedJwtVerificationOpts.java (hunk 1 above)
Comment (IngridPuppet) [reply src/main/java/de/adorsys/sdjwt/IssuerSignedJwtVerificationOpts.java:39]: I think it is okay to leave the risk of a null pointer exception to the caller. As of now, the exception can come from the class itself.

File: src/test/java/de/adorsys/sdjwt/TimeClaimVerifierTest.java (hunk 2 above)
Comment (IngridPuppet) [reply src/test/java/de/adorsys/sdjwt/TimeClaimVerifierTest.java]: I agree. Thank you.

File: src/test/java/de/adorsys/sdjwt/TimeClaimVerifierTest.java (hunk 3 above)
Comment (IngridPuppet) [reply src/test/java/de/adorsys/sdjwt/TimeClaimVerifierTest.java:43]: As a checked exception `SdJwtVerificationException` that can be thrown from the methods called, Java requires this explicit syntax in the case a try/catch block is not used.

File: src/test/java/de/adorsys/sdjwt/TimeClaimVerifierTest.java (hunk 4 above)
Comment (IngridPuppet) [reply src/test/java/de/adorsys/sdjwt/TimeClaimVerifierTest.java:71]: As a checked exception `SdJwtVerificationException` that can be thrown from the methods called, Java requires this explicit syntax in the case a try/catch block is not used.

File: src/main/java/de/adorsys/sdjwt/TimeClaimVerificationOpts.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    public static final int DEFAULT_LEEWAY_SECONDS = 10;
+
+    private final boolean validateExpirationClaim;
+    private final boolean validateNotBeforeClaim;
+    private final int leewaySeconds;
+
+    public TimeClaimVerificationOpts(
+            boolean validateExpirationClaim,
+            boolean validateNotBeforeClaim,
+            int leewaySeconds) {
```
Comment (Ogenbertrand) [inline src/main/java/de/adorsys/sdjwt/TimeClaimVerificationOpts.java:22]: It would be prudent to validate the input parameters in the constructor to ensure they are within acceptable ranges (e.g., leewaySeconds should not be negative).
 
 ```java
 if (leewaySeconds < 0) {
    throw new IllegalArgumentException("Leeway seconds cannot be negative");
}
```

File: src/main/java/de/adorsys/sdjwt/TimeClaimVerificationOpts.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+public class TimeClaimVerificationOpts {
+
+    /**
+     * Tolerance window to account for clock skew when checking time claims
+     */
+    public static final int DEFAULT_LEEWAY_SECONDS = 10;
+
+    private final boolean validateExpirationClaim;
+    private final boolean validateNotBeforeClaim;
+    private final int leewaySeconds;
```
Comment (Ogenbertrand) [inline src/main/java/de/adorsys/sdjwt/TimeClaimVerificationOpts.java:17]: The class has a JavaDoc comment, but it could be expanded to include descriptions for the constructor and methods. This would help other developers understand the purpose and usage of each method more clearly. What do you think ?

File: src/main/java/de/adorsys/sdjwt/TimeClaimVerifier.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+import com.fasterxml.jackson.databind.JsonNode;
+import de.adorsys.sdjwt.exception.SdJwtVerificationException;
+
+import java.time.Instant;
+
+/**
+ * Module for checking the validity of JWT time claims
+ */
+public class TimeClaimVerifier {
+
```
Comment (Ogenbertrand) [inline src/main/java/de/adorsys/sdjwt/TimeClaimVerifier.java:12]: The constructor does not validate the leewaySeconds still. It would be wise to ensure that it is non-negative. for example;

```java
public TimeClaimVerifier(int leewaySeconds) {
    if (leewaySeconds < 0) {
        throw new IllegalArgumentException("Leeway seconds cannot be negative");
    }
    this.leewaySeconds = leewaySeconds;
}
```

File: src/main/java/de/adorsys/sdjwt/TimeClaimVerifier.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    /**
+     * Tolerance window to account for clock skew
+     */
+    private final int leewaySeconds;
```
Comment (Ogenbertrand) [inline src/main/java/de/adorsys/sdjwt/TimeClaimVerifier.java:19]: Consider defining a constant for the claim names ("exp" and "nbf") to avoid magic strings and improve maintainability. And also adding JavaDoc comments for the methods to improve readability

File: src/test/java/de/adorsys/sdjwt/TimeClaimVerifierTest.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+    // Test for verifyNotBeforeClaim (edge case: valid exactly at current time with leeway)
+    @Test
+    public void testVerifyNotBeforeClaimEdge() throws SdJwtVerificationException {
+        ObjectNode payload = SdJwtUtils.mapper.createObjectNode();
+        payload.put("nbf", CURRENT_TIMESTAMP + 59); // 59 seconds in the future, within the 60 second leeway
+
+        // No exception expected for JWT becoming valid within leeway
+        timeClaimVerifier.verifyNotBeforeClaim(payload);
+    }
```
Comment (Ogenbertrand) [inline src/test/java/de/adorsys/sdjwt/TimeClaimVerifierTest.java:86]: This test class covers a good range of scenarios, including expired claims, valid claims, and edge cases. But again adding tests for invalid inputs, such as null payloads or missing claims, to ensure robustness.

**Review by Ogenbertrand — CHANGES_REQUESTED** [review]

File: src/main/java/de/adorsys/sdjwt/TimeClaimVerificationOpts.java (hunk 5 above)
Comment (IngridPuppet) [reply src/main/java/de/adorsys/sdjwt/TimeClaimVerificationOpts.java:22]: Agree. I'm adding the check to the `TimeClaimVerifier` constructor directly, where the value actually matters.

File: src/main/java/de/adorsys/sdjwt/TimeClaimVerificationOpts.java (hunk 6 above)
Comment (IngridPuppet) [reply src/main/java/de/adorsys/sdjwt/TimeClaimVerificationOpts.java:17]: I think that the method names are self explanatory enough. There is no complex logic going on here.

File: src/main/java/de/adorsys/sdjwt/TimeClaimVerifier.java (hunk 7 above)
Comment (IngridPuppet) [reply src/main/java/de/adorsys/sdjwt/TimeClaimVerifier.java:12]: Thanks, added.

File: src/main/java/de/adorsys/sdjwt/TimeClaimVerifier.java (hunk 8 above)
Comment (IngridPuppet) [reply src/main/java/de/adorsys/sdjwt/TimeClaimVerifier.java:19]: Thanks. I defined constants for the claim names. Note that the methods of this class already have Java docs.

File: src/test/java/de/adorsys/sdjwt/TimeClaimVerifierTest.java (hunk 9 above)
Comment (IngridPuppet) [reply src/test/java/de/adorsys/sdjwt/TimeClaimVerifierTest.java:86]: Thank you for your suggestion. I added a `requireNonNull` check to guarantee readable exceptions on null payloads. Note that missing claims are already covered by the tests of the corresponding layer.

**Review by Ogenbertrand — APPROVED** [review]
